File: DemoServer/AdditionalLanguages/python/facetedSearch/facetsAggregations/facetsAggregations.py
from models import Product
from demo_example import Example, RunParamsBase

#region Usings
from ravendb import AbstractIndexCreationTask, Facet, FacetAggregationField, RangeFacet
from ravendb.documents.queries.facets.misc import FacetAggregation
#endregion

class RunParams(RunParamsBase):
    def __init__(self, range_1: int = None, range_2: int = None, range_3: int = None):
        self.range_1 = range_1
        self.range_2 = range_2
        self.range_3 = range_3

#region Demo
#region Step_1
class Products_ByCategoryPriceAndUnits(AbstractIndexCreationTask):
    def __init__(self):
        super().__init__()
        self.map = (
            "docs.Products.Select(product => new {"
            '    CategoryName = (this.LoadDocument(product.Category, "Categories")).Name,'
            "    PricePerUnit = product.PricePerUnit,"
            "    UnitsInStock = product.UnitsInStock"
            "})"
        )
#endregion
#endregion

class FacetsAggregations(Example):
    def run(self, run_params: RunParams):
        facets_results = []
        range1 = run_params.range_1 or 25.0
        range2 = run_params.range_2 or 50.0
        range3 = run_params.range_3 or 100.0
        Products_ByCategoryPriceAndUnits().execute(self.document_store_holder.store())
        #region Demo
        #region Step_2
        facet = Facet()
        facet.field_name = "CategoryName"

        price_per_unit_aggregation_field = FacetAggregationField()
        price_per_unit_aggregation_field.name = "PricePerUnit"

        units_in_stock_aggregation_field = FacetAggregationField()
        units_in_stock_aggregation_field.name = "UnitsInStock"

        facet.aggregations.update({FacetAggregation.AVERAGE: {price_per_unit_aggregation_field}})
        facet.aggregations.update({FacetAggregation.SUM: {units_in_stock_aggregation_field}})
        facet.aggregations.update({FacetAggregation.MAX: {price_per_unit_aggregation_field}})

        fields = {price_per_unit_aggregation_field, units_in_stock_aggregation_field}

        facet.aggregations.update({FacetAggregation.MIN: fields})
        #endregion

        #region Step_3
        range_facet = RangeFacet()
        range_facet.ranges = [
            f"PricePerUnit < {range1}",
            f"PricePerUnit >= {range1} and PricePerUnit < {range2}",
            f"PricePerUnit >= {range2} and PricePerUnit < {range3}",
            f"PricePerUnit > {range3}",
        ]

        range_facet.aggregations.update({FacetAggregation.AVERAGE: {price_per_unit_aggregation_field}})
        range_facet.aggregations.update({FacetAggregation.SUM: {units_in_stock_aggregation_field}})
        range_facet.aggregations.update({FacetAggregation.MAX: {price_per_unit_aggregation_field}})
        range_facet.aggregations.update({FacetAggregation.MIN: fields})
        #endregion

        #region Step_4
        facets = [facet, range_facet]
        #endregion

        with self.document_store_holder.store().open_session() as session:
            #region Step_5
            query_results = (
                session.query_index_type(Products_ByCategoryPriceAndUnits, Product)
                    .aggregate_by_facets(*facets)
                    .execute()
            )
            #endregion
        #endregion

        return query_results

File: DemoServer/AdditionalLanguages/php/facetedSearch/facetsAggregations/FacetsAggregations.php
<?php

namespace RavenDB\Demo\facetedSearch\facetsAggregations;

//region Usings
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Documents\Queries\Facets\AggregationArray;
use RavenDB\Documents\Queries\Facets\Facet;
use RavenDB\Documents\Queries\Facets\FacetAggregation;
use RavenDB\Documents\Queries\Facets\FacetAggregationField;
use RavenDB\Documents\Queries\Facets\FacetAggregationFieldSet;
use RavenDB\Documents\Queries\Facets\FacetResultArray;
use RavenDB\Documents\Queries\Facets\RangeFacet;
//endregion

use RavenDB\Demo\common\DocumentStoreHolder;
use RavenDB\Demo\common\models\Product;

//region demo
//region Step_1
class Products_ByCategoryPriceAndUnits extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->map = "docs.Products.Select(product => new {" .
            "    CategoryName = (this.LoadDocument(product.Category, \"Categories\")).Name," .
            "    PricePerUnit = product.PricePerUnit," .
            "    UnitsInStock = product.UnitsInStock" .
            "})";
    }
}
//endregion
//endregion


class FacetsAggregations
{
    public function __invoke(RunParams $runParams): FacetResultArray
    {
        $range1 = $runParams->getRange1() ?? 25.0;
        $range2 = $runParams->getRange2() ?? 50.0;
        $range3 = $runParams->getRange3() ?? 100.0;

        //region Demo
        //region Step_2
        $facet = new Facet();
        $facet->setFieldName("CategoryName");

        $pricePerUnitAggregationField = new FacetAggregationField();
        $pricePerUnitAggregationField->setName("PricePerUnit");

        $unitsInStockAggregationField = new FacetAggregationField();
        $unitsInStockAggregationField->setName("UnitsInStock");

        $facet->getAggregations()->set(FacetAggregation::average(), FacetAggregationFieldSet::fromArray([ $pricePerUnitAggregationField ]));
        $facet->getAggregations()->set(FacetAggregation::sum(), FacetAggregationFieldSet::fromArray([ $unitsInStockAggregationField ]));
        $facet->getAggregations()->set(FacetAggregation::max(), FacetAggregationFieldSet::fromArray([ $pricePerUnitAggregationField ]));

        $fields = [];
        $fields[] = $pricePerUnitAggregationField;
        $fields[] = $unitsInStockAggregationField;

        $facet->getAggregations()->set(FacetAggregation::min(), FacetAggregationFieldSet::fromArray($fields));
        //endregion

        //region Step_3
        $rangeFacet = new RangeFacet();
        $rangeFacet->setRanges([
                "PricePerUnit < " . $range1,
                "PricePerUnit >= " . $range1 . " and PricePerUnit < " . $range2,
                "PricePerUnit >= " . $range2 . " and PricePerUnit < " . $range3,
                "PricePerUnit >= " . $range3
            ]);

        $aggregations = new AggregationArray();

        $aggregations->offsetSet(FacetAggregation::AVERAGE, FacetAggregationFieldSet::fromArray([ $pricePerUnitAggregationField ]));
        $aggregations->offsetSet(FacetAggregation::SUM, FacetAggregationFieldSet::fromArray([ $unitsInStockAggregationField ]));
        $aggregations->offsetSet(FacetAggregation::MAX, FacetAggregationFieldSet::fromArray([ $pricePerUnitAggregationField ]));
        $aggregations->offsetSet(FacetAggregation::MIN, FacetAggregationFieldSet::fromArray($fields));

        $rangeFacet->setAggregations($aggregations);

        //endregion

        //region Step_4
        $facets = [
            $facet,
            $rangeFacet
        ];
        //endregion

        $queryResults = [];

        $session = DocumentStoreHolder::getStore()->openSession();
        try {
            //region Step_5
            $queryResults = $session->query(Product::class, Products_ByCategoryPriceAndUnits::class)
                ->aggregateBy($facets)
                ->execute();
            //endregion
        } finally {
            $session->close();
        }
        //endregion

        return $queryResults;
    }
}
